### PR TITLE
Webpack-friendly images for Storybook

### PR DESCRIPTION
**Pull request by bencodeorg**
These are some examples of how we can get relevant assets to load in Storybook without including all of our static asset directories.

As one example, before this change, the `ShareAllowedDialog` default thumbnail in [our public storybook](https://code-dot-org.github.io/cdo-styleguide/?path=/story/sharealloweddialog--abusive) is missing, and after, it is not (screenshot from a local build):

| Before      | After |
| ----------- | ----------- |
| image | image |

(preemptive apologies, as my understanding is a little fuzzy here / I may be using the wrong terms, but hopefully the general idea is clear)

As I understand it, using this syntax (rather than a hard coded path like `/blockly/media/...`) results in Webpack being smart and replacing references to the image path with the webpacked location of the image. Because we use webpack config to deploy storybook, webpack figures out which images need to be included in the deploy and includes them.

This is a pattern we use in some places already, so some images already work in Storybook. See examples [like this one](https://github.com/code-dot-org/cdo-styleguide/blob/gh-pages/applabcreateprojectwpcf448e2676e480e4b6e76d1f2a789375.png) in the top level cdo-styleguide repo ([here's the relevant code](https://github.com/code-dot-org/code-dot-org/blob/94b4fd20149c23e2cb8fc07eaf64af016554fab3/apps/src/templates/VerticalImageResourceCard.jsx#L57) in our code base). After this change, we should see a similar file for `project_default.png` and others included in the deploy.

Here are some [relevant docs](https://webpack.js.org/guides/asset-management/#loading-images), and the key bit:

> Now, when you `import MyImage from './my-image.png'`, that image will be processed and added to your output directory and the MyImage variable will contain the final url of that image after processing. 

Its worth noting that as far as I can tell, assets like these didn't load when Storybook last was deployed 5 years ago, so I think its ok if we try to move in this direction over time.

## Testing story

I tested building storybook locally, and saw that a file like `project_defaultwp7836f8ff25cbef1ff45bd0eaf4e1977e.png` was included in the top level directory. I viewed the generated `index.html` for Storybook and saw the "after" screenshot above. I also looked at the share dialog in a new Applab project on Code Studio, and it looked good as well. All of the other changes apply to Storybook only, and I was able to view the images successfully in a local build.

### CODE REVIEW

File: apps/src/code-studio/components/ShareAllowedDialog.jsx [hunk 1]
```diff
@@ -22,6 +22,7 @@ import QRCode from 'qrcode.react';
 import copyToClipboard from '@cdo/apps/util/copyToClipboard';
 import FontAwesome from '@cdo/apps/templates/FontAwesome';
 import Button from '../../templates/Button';
+import defaultThumbnail from '@cdo/static/projects/project_default.png';
```
Comment (fisher-alice) [inline apps/src/code-studio/components/ShareAllowedDialog.jsx:25]: Curious why you updated image in this file which is not a .storybook file. Did you mean to update `Dialog.story.jsx`?

Comment (bencodeorg) [reply apps/src/code-studio/components/ShareAllowedDialog.jsx:25]: No, this was intentional -- this style works in any context (ie, story or not), it just plays particularly nicely with Storybook, which relies on this style to resolve the paths of assets like images.

The other .`story.jsx` files were the ones where I could find files `/blockly/media` URLs directly in them (ie, by searching for that string in files with the story path suffix). I'm sure there are other places where we could update and remove broken images from Storybook, but I didn't want to go through each component individually to find broken images (a lot of components look pretty dated / maybe don't belong in Storybook, so didn't feel 100% useful to do so).

**Review by sanchitmalhotra126 — APPROVED** [review]
Interesting! I'm also pretty fuzzy on these static paths so this was helpful to walk through.

One thing I noticed is that in our Gruntfile, we have a mapping for `blockly/media` in our Karma config ([here](https://github.com/code-dot-org/code-dot-org/blob/57ccf70b6211b422ecb830016a463a7db6049220/apps/Gruntfile.js#L505)). I'd be curious to know if we could somehow create a similar mapping for Storybook, maybe we wouldn't need to change the paths? That being said, relying on Webpack to fetch the right image paths seems to be a better solution regardless. Sounds like something we should just be doing more consistently across the codebase. For example I also just noticed we happen to use both forms in Music Lab and things seem to work fine ([blockly/media](https://github.com/code-dot-org/code-dot-org/blob/57ccf70b6211b422ecb830016a463a7db6049220/apps/src/music/views/SoundsPanel.jsx#L85) vs [@cdo/static](https://github.com/code-dot-org/code-dot-org/blob/57ccf70b6211b422ecb830016a463a7db6049220/apps/src/music/utils/Loader.ts#L23))

**Comment by bencodeorg** [comment]
> One thing I noticed is that in our Gruntfile, we have a mapping for `blockly/media` in our Karma config ([here](https://github.com/code-dot-org/code-dot-org/blob/57ccf70b6211b422ecb830016a463a7db6049220/apps/Gruntfile.js#L505)). I'd be curious to know if we could somehow create a similar mapping for Storybook, maybe we wouldn't need to change the paths? That being said, relying on Webpack to fetch the right image paths seems to be a better solution regardless. 

That's an interesting idea -- taking a look at the Karma docs (although I think this is the wrong version), it looks like this really lets you proxy (almost like an alias?) a path:

http://karma-runner.github.io/6.4/config/configuration-file.html#proxies

I think if we could configure something like substituting absolute paths from another environment (eg, studio.code.org/blockly/media/...) for the relative paths, this could work? But I don't see a way to do that in Storybook -- it seems like you can just configure which directories are included in a built version:

https://storybook.js.org/docs/react/configure/images-and-assets

LMK if there's another angle here you were envisioning.

> Sounds like something we should just be doing more consistently across the codebase. For example I also just noticed we happen to use both forms in Music Lab and things seem to work fine ([blockly/media](https://github.com/code-dot-org/code-dot-org/blob/57ccf70b6211b422ecb830016a463a7db6049220/apps/src/music/views/SoundsPanel.jsx#L85) vs [@cdo/static](https://github.com/code-dot-org/code-dot-org/blob/57ccf70b6211b422ecb830016a463a7db6049220/apps/src/music/utils/Loader.ts#L23))

Yeah, both work fine. I think the Webpack-friendly approach is better as it allows Webpack to manage assets for us, which I think it what it's for. [This is an old doc](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/docs/build.md), but I think this section is relevant:

> _The capabilities of the rails asset pipeline overlap significantly with those of webpack. In attempt to keep things simple, our current design is to do as much asset management work as possible in webpack and as little as possible in rails._

**Pull request merged**